packages/melody-runtime/__tests__/AsyncComponentSpec.js

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+import {
+    patch,
+    elementOpen,
+    text,
+    component,
+    elementClose,
+    enqueueComponent,
+    flush,
+} from 'melody-idom';
+import { AsyncComponent } from '../src';
+
+describe('AsyncComponent', () => {
+    let unmounted = false;
+    let notified = false;
+    let el = null;
+
+    beforeEach(function() {
+        unmounted = false;
+        notified = false;
+        el = document.createElement('div');
+    });
+
+    class Component {
+        constructor() {
+            this.el = null;
+            this.refs = Object.create(null);
+            this.props = null;
+        }
+
+        apply(props) {
+            this.props = props;
+            enqueueComponent(this);
+        }
+
+        notify() {
+            notified = true;
+        }
+
+        componentWillUnmount() {
+            unmounted = true;
+        }
+
+        render() {
+            elementOpen('div');
+            text(this.props.text);
+            elementClose('div');
+        }
+    }
+    it('should render a promised component', async function() {
+        const template = data => {
+            component(AsyncComponent, 'test', {
+                promisedComponent: () =>
+                    Promise.resolve({ default: Component }),
+                whileLoading: () => {
+                    elementOpen('b');
+                    text('Loading...');
+                    elementClose('b');
+                },
+                onError: error => {
+                    elementOpen('strong');
+                    text(error);
+                    elementClose('strong');
+                },
+                data,
+            });
+            component(AsyncComponent, 'test_failure', {
+                promisedComponent: () =>
+                    Promise.reject('Network connection issue'),
+                whileLoading: () => {
+                    elementOpen('b');
+                    text('Loading...');
+                    elementClose('b');
+                },
+                onError: error => {
+                    elementOpen('strong');
+                    text(error);
+                    elementClose('strong');
+                },
+                data,
+            });
+        };
+        patch(el, template, { text: 'Hello' });
+        expect(el.innerHTML).toEqual(
+            '<m-placeholder></m-placeholder><m-placeholder></m-placeholder>'
+        );
+        run(2);
+        expect(el.innerHTML).toEqual('<b>Loading...</b><b>Loading...</b>');
+        await Promise.resolve();
+        run(2);
+        expect(el.innerHTML).toEqual(
+            '<div>Hello</div><strong>Network connection issue</strong>'
+        );
+
+        patch(el, template, { text: 'Foo' });
+        run();
+        expect(el.innerHTML).toEqual(
+            '<div>Foo</div><strong>Network connection issue</strong>'
+        );
+
+        expect(unmounted).toEqual(false);
+    });
+});
+
+function run(rounds = 1) {
```

Comment:
Would be nice to have some more tests, that check proper cancelling, unmounting (eg. with conditions):

```twig
// What happens when qux changes multiple times, while the component is loading?
{% if qux %}
  {% mount async 'foo.twig' as 'bar' with {foo: 'bar'} delay placeholder by 500ms %}
    Loading...
  {% endmount %}
{% else %}
  {% mount async 'bar.twig' as 'foo' with {foo: 'bar'} delay placeholder by 500ms %}
    Loading...
  {% endmount %}
{% endif %}
```

Reply:
Agreed. More tests are definitely needed.

Reply:
Added some more